### ISSUE TITLE
To see memory probabilities

### ISSUE THREAD

**Issue by joecheriross** [domluna/memn2n#16]
I am trying to see the memory slot probabilities(probabilities associated with different sentences) for a particular query. Is there a way to visualize them ? Please help. 

Thanks,
Joe

**Comment by domluna** [comment]
I think it would be cool, thought about it in the past but I can't guarantee I'll get around to it. If you come up with something feel free to send a pull request!

**Comment by joecheriross** [comment]
A few questions to make sure my understanding is correct. 
During testing, providing stories and questions. 

- In  memn2n.py
`probs = tf.nn.softmax(dotted)` 
The `probs` here get the probability values corresponding to each memory slot(sentence in the story)  for the given query in each hop right ?
If there are 3 sentences in the story, then `probs` will have 3 probability values

- Finding a way to print `probs` and `stories` should help to get an idea on how the sentences are picked for answering a question ?

**Comment by domluna** [comment]
That's correct. Maybe
https://www.tensorflow.org/versions/r0.10/api_docs/python/control_flow_ops.html#Print
could help in this regard.

**Comment by joecheriross** [comment]
I tried out some modification (in memn2n.py) to get the memory probabilities (attentions).  The last `probs` obtained, is assigned to a variable `self.mem_probs` after the hops loop in _inference(). This is returned by the predict_proba() function. 

But each time when I run, I get different values for the memory probabilities for the same test instances. I was trying to get  accuracy based on attentions. But this gives different accuracy each time. Can you please see if I am doing something wrong ? Please let me know if something is not clear. 

modification in memn2n.py
====================
  ```
            probs= None #added by joe
            for _ in range(self._hops):
                m_emb = tf.nn.embedding_lookup(self.A, stories)
                m = tf.reduce_sum(m_emb * self._encoding, 2) + self.TA
                # hack to get around no reduce_dot
                u_temp = tf.transpose(tf.expand_dims(u[-1], -1), [0, 2, 1])
                dotted = tf.reduce_sum(m * u_temp, 2)

                # Calculate probabilities
                probs = tf.nn.softmax(dotted)
                #pr= tf.Print(probs, [probs], message="probs: ", summarize=1000000000)
                #self.image_summary_t= tf.image_summary("probs", probs)

                probs_temp = tf.transpose(tf.expand_dims(probs, -1), [0, 2, 1])
                c_temp = tf.transpose(m, [0, 2, 1])
                o_k = tf.reduce_sum(c_temp * probs_temp, 2)


                u_k = tf.matmul(u[-1], self.H) + o_k
                # nonlinearity
                if self._nonlin:
                    u_k = nonlin(u_k)

                u.append(u_k)

            self.mem_probs= probs #added by joe

```

modification in predict_proba
======================
`return self._sess.run([self.predict_proba_op, self.mem_probs], feed_dict=feed_dict)`

**Comment by domluna** [comment]
Does `#pr= tf.Print(probs, [probs], message="probs: ", summarize=1000000000)` give the same probs as `self.mem_probs`?

**Comment by joecheriross** [comment]
Yes. I checked that as well, to make sure that I am not doing something wrong. 
It also gives the same value.
